File: picnic_app/lib/models/vote/artist.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:picnic_app/models/vote/artist_group.dart';

part '../../generated/models/vote/artist.freezed.dart';

part '../../generated/models/vote/artist.g.dart';

@freezed
class ArtistModel with _$ArtistModel {
  const ArtistModel._();

  const factory ArtistModel({
    required int id,
    required Map<String, dynamic> name,
    int? yy,
    int? mm,
    int? dd,
    DateTime? birth_date,
    String? gender,
    ArtistGroupModel? artist_group,
    String? image,
    DateTime? created_at,
    DateTime? updated_at,
    DateTime? deleted_at,
    bool? isBookmarked,
  }) = _ArtistModel;

  factory ArtistModel.fromJson(Map<String, dynamic> json) =>
      _$ArtistModelFromJson(json);

  DateTime? get birthDate {
    if (birth_date != null) return birth_date;
    if (yy != null && mm != null && dd != null) {
      try {
        return DateTime(yy!, mm!, dd!);
      } catch (e) {
        return null;
      }
    }
    return null;
  }

  String? get formattedBirthDate {
    final date = birthDate;
    if (date == null) return null;
    return '${date.year}년 ${date.month}월 ${date.day}일';
  }

  String? get formattedName {
    if (name.containsKey('ko')) {
      return name['ko'];
    } else if (name.containsKey('en')) {
      return name['en'];
    } else if (name.isNotEmpty) {
      return name.values.first;
    }
    return null;
  }

  bool get isDeleted => deleted_at != null;
}

File: packages/web/admin/picnic-admin/lib/types/artist.ts
// // DB 스키마 기반 아티스트 인터페이스
// export interface ArtistSchema {
//   id: number;
//   created_at: string;
//   updated_at: string;
//   deleted_at: string | null;
//   name: Record<string, string>; // JSON 타입
//   yy: number | null;
//   mm: number | null;
//   dd: number | null;
//   gender: string | null;
//   artist_group_id: number;
//   image: string | null;
//   birth_date: string | null;
//   debut_yy: number | null;
//   debut_mm: number | null;
//   debut_dd: number | null;
//   artist_group?: ArtistGroupSchema;
// }

// // DB 스키마 기반 아티스트 그룹 인터페이스
// export interface ArtistGroupSchema {
//   id: number;
//   created_at: string;
//   updated_at: string;
//   deleted_at: string | null;
//   name: Record<string, string>; // JSON 타입
//   image: string;
//   debut_yy: number | null;
//   debut_mm: number | null;
//   debut_dd: number | null;
//   artists?: ArtistSchema[];
// }

// 애플리케이션에서 사용할 통합 아티스트 인터페이스
export interface Artist {
  id: string | number;
  name?: Record<string, string>;
  image?: string | null;
  birth_date?: string | null;
  yy?: number | null;
  mm?: number | null;
  dd?: number | null;
  gender?: string | null;
  artist_group_id?: number;
  group_id?: number;
  created_at?: string;
  updated_at?: string;
  deleted_at?: string | null;
  debut_date?: string | null;
  debut_yy?: number | null;
  debut_mm?: number | null;
  debut_dd?: number | null;
  artist_group?: ArtistGroup;
}

// 애플리케이션에서 사용할 통합 아티스트 그룹 인터페이스
export interface ArtistGroup {
  id: string;
  name: {
    ko: string;
    en: string;
    ja: string;
    zh: string;
    id: string;
  };
  image?: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
  debut_date?: string | null;
  debut_yy?: number | null;
  debut_mm?: number | null;
  debut_dd?: number | null;
  artists?: Artist[];
}
